server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala

```diff
@@ -268,6 +296,59 @@ class InteractiveSession(
     }
   }
 
+  private def datanucleusJars(livyConf: LivyConf): Seq[String] = {
+    if (sys.env.getOrElse("LIVY_INTEGRATION_TEST", "false").toBoolean) {
+      // datanucleus jars has already been in classpath in integration test
+      Seq.empty
+    } else {
+      val sparkHome = livyConf.sparkHome().get
+      val libdir =
+        if (new File(sparkHome, "RELEASE").isFile) {
+          new File(sparkHome, "lib")
+        } else {
+          new File(sparkHome, "lib_managed/jars")
+        }
+      val jars = if (!libdir.isDirectory) {
+          Seq.empty[String]
+        } else {
+          libdir.listFiles().filter(_.getName.startsWith("datanucleus-"))
+            .map(_.getAbsolutePath).toSeq
+        }
+      if (jars.isEmpty) {
+        warn("datanucleus jars can not be found")
+      }
+      jars
+    }
+  }
+
+  /**
+   * Look for hive-site.xml (for now just ignore spark.files defined in spark-defaults.conf)
+   * 1. First look for hive-site.xml in user request
+   * 2. Then look for that under classpath
+   * 3. Finally look for that under SPARK_HOME/conf
+   * @param livyConf
+   * @return  (hive-site.xml path, whether it is provided by user)
+   */
+  private def hiveSiteFile(sparkFiles: Array[String],
+                           livyConf: LivyConf): (Option[File], Boolean) = {
+    if (sparkFiles.exists(_.split("/").last == "hive-site.xml")) {
+      (None, true)
+    } else {
+      val hiveSiteURL = getClass.getResource("/hive-site.xml")
+      if (hiveSiteURL != null && hiveSiteURL.getProtocol == "file") {
+        return (Some(new File(hiveSiteURL.toURI)), false)
+      }
+      // only check SPARK_HOME/conf when SPARK_CONF_DIR is not defined
+      if (!sys.env.contains("SPARK_CONF_DIR")) {
```

Comment:
I think this is actually a bug in `bin/livy-server`. If `SPARK_CONF_DIR` is not defined it should be adding `$SPARK_HOME/conf` to the classpath in its place, otherwise its behavior is inconsistent.

Another possibility is to not add either to the classpath, although I'm not sure what other code that would affect.